logprep/processor/generic_adder/processor.py

```diff
@@ -43,15 +56,131 @@ def __init__(self, name: str, skipped_fields: List[str]):
+        self._db_connector = MySQLConnector(sql_config, logger) if sql_config else None
+
+        if GenericAdder.db_table is None:
+            GenericAdder.db_table = self._db_connector.get_data() if self._db_connector else None
+        self._db_table = GenericAdder.db_table
+
+    def _apply_rules(self, event: dict, rule: GenericAdderRule):
+        """Apply a matching generic adder rule to the event.
+         Add fields and values to the event according to the rules it matches for.
+        Additions can come from the rule definition, from a file or from a SQL table.
+        The SQL table is initially loaded from the database and then reloaded if it changes.
+        At first it checks if a SQL table exists and if it will be used. If it does, it adds all
+        values from a matching row in the table to the event. To determine if a row matches, a
+        pattern is used on a defined value of the event to extract a subvalue that is then matched
+        against a value in a defined column of the SQL table. A dotted path prefix can be applied to
+        add the new fields into a shared nested location.
+        If no table exists, fields defined withing the rule itself or in a rule file are being added
+        to the event.
+        Parameters
+        ----------
+        event : dict
+           Name of the event to add keys and values to.
+        rule : GenericAdderRule
+           A matching generic adder rule.
+        Raises
+        ------
+        DuplicationError
+            Raises if an addition would overwrite an existing field or value.
+        """
+
+        if self._db_connector and self._db_connector.check_change():
+            self._db_table = self._db_connector.get_data()
+
+        conflicting_fields = []
+
+        # Either add items from a sql db table or from the rule definition and/or a file
+        if rule.db_target and self._db_table:
+            # Create items to add from a db table
+            items_to_add = []
+            if rule.db_pattern:
+                # Get the sub part of the value from the event using a regex pattern
+                value_to_check_in_db = self._get_dotted_field_value(event, rule.db_target)
+                match_with_value_in_db = rule.db_pattern.match(value_to_check_in_db)
+                if match_with_value_in_db:
+                    # Get values to add from db table using the sub part
+                    value_to_map = match_with_value_in_db.group(1).upper()
+                    to_add_from_db = self._db_table.get(value_to_map, [])
+                    for item in to_add_from_db:
+                        if rule.db_destination_prefix:
+                            if not item[0].startswith(rule.db_destination_prefix):
+                                item[0] = f"{rule.db_destination_prefix}.{item[0]}"
+                        items_to_add.append(item)
+        else:
+            # Use items from rule definition and/or file
+            items_to_add = rule.add.items()
```

Comment:
too many nested blocks. please refactor to ensure readability.
Hint: handle special cases first and negate if statements to avoid nesting

Reply:
I've changed it.